torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp

```diff
@@ -0,0 +1,350 @@
+    }
+  }
+}
+
+void inlineScopeBlocks(Block* b) {
+  for (auto n_itr = b->nodes().begin(); n_itr != b->nodes().end();) {
+    Node* n = *n_itr++;
+    for (Block* sub_b : n->blocks()) {
+      inlineScopeBlocks(sub_b);
+    }
+    if (n->kind() == prim::FakeScopeBlock) {
+      // Convert the block to a graph so we can inline it
+      auto graph = std::make_shared<Graph>();
+      std::unordered_map<Value*, Value*> remaps;
+      graph->block()->cloneFrom(n->blocks()[0], [&](Value* v) {
+        remaps[v] = graph->block()->addInput()->copyMetadata(v);
+        n->addInput(v);
+        return remaps[v];
+      });
+
+      inlineCallTo(n, *graph);
+    }
+  }
+}
+
+void runCleanupPasses(const std::shared_ptr<Graph>& g) {
+  if (script::getInlineEverythingMode()) {
+    Inline(*g);
+  }
+  LowerSimpleTuples(g);
+  EliminateDeadCode(g);
+  LintGraph(g);
+}
+
+void runCleanupPasses(script::Module* m) {
+  auto methods = m->get_methods();
+  for (auto module : m->get_modules()) {
+    runCleanupPasses(&module.module);
+  }
+  for (auto& method : methods) {
+    runCleanupPasses(method.graph());
+  }
+}
+
+void FixupTraceScopeBlocks(
+    std::shared_ptr<Graph>& graph,
+    script::Module* self) {
+  // Gather level information about blocks in the graph.
+  // Assign each block a number, with graph->block() gettng
+  // 0 and each sub-block gets a higher number based on how
+  // deeply nested it is.
+  BlockInfoMap block_info;
+  inspectBlocks(graph->block(), block_info);
+  // Iterate through all the nodes in program order and--for each use--
+  // if the Value referenced is not in a scope that dominates the node,
+  // add block and Node outputs to lift it into a scope in which
+  // it dominates the Use.
+  RemappingTable table;
+  makeDefsDominateUses(graph->block(), table, block_info);
+  // For all blocks except graph->block(), convert multiple block
+  // returns to a TupleConstruct. This is required for turning the
+  // blocks into Methods. (and in the case that self is nullptr,
+  // it is required to properly inline the blocks).
+  convertReturnsToTuples(graph->block());
+  if (!self) {
+    // We have no Module, so we're just going to inline everything.
+    // This should give us a totally flag graph.
+    inlineScopeBlocks(graph->block());
+    runCleanupPasses(graph);
+  } else {
+    // Lambda lift Values (i.e. add Graph inputs for the purpose of
+    // referencing values that dominate the block) and convert
+    // the block to a Graph. blocks()[0] on each FakeScopeBlock then
+    // appears as a Graph attribute attr::Subgraph
+    lambdaLiftBlocksAndConvertToGraph(graph->block());
+    // Register the attr::Subgraph Graph values as Functions in the
+    // class compilation unit and register that Function as a method
+    // on the corresponding Module in the Module hierarchy. Note that we
+    // unique the methods by naming them forward, forward1, forward2...
+    createMethodCalls(graph, *self, {"__module"});
```

Comment:
Can you make an argument for the algorithmic complexity of this whole process? I want to make sure it is not N^2 as the module depth grows, but I cannot tell by inspection. Most of these passes are recursive so it is not clear what happens when we repeatedly apply them.